ssl/ssl_lib.c

```diff
@@ -566,10 +566,12 @@ int SSL_clear(SSL *s)
 
     /*
      * Check to see if we were changed into a different method, if so, revert
-     * back if we are not doing session-id reuse.
+     * back. We always do this in TLSv1.3. Below that we only do it if we are
+     * not doing session-id reuse.
      */
-    if (!ossl_statem_get_in_handshake(s) && (s->session == NULL)
-        && (s->method != s->ctx->method)) {
+    if (s->method != s->ctx->method
+            && (SSL_IS_TLS13(s)
+                || (!ossl_statem_get_in_handshake(s) && s->session == NULL))) {
```

Comment:
I think you can remove this condition altogether and just do the `s->method != s->ctx->method`. The session-id mess as a remnant of `SSL_set_session` locking you to that version. The check exists so you do not accidentally undo the behavior remove here:
https://github.com/openssl/openssl/commit/ccae4a1582efcad311d095a8e6832b2b67d5ed05#diff-1d72223f932d41f617c50222791649ec

But since that behavior is removed, `SSL_clear` should follow suit and not special-case session behavior.